src/nvme/ioctl.c

```diff
@@ -1617,10 +1669,26 @@ int nvme_io(struct nvme_io_args *args, __u8 opcode)
 		.timeout_ms	= args->timeout,
 	};
 
-	if (args->args_size < sizeof(*args)) {
+	if (args->args_size < sizeof_args(struct nvme_io_args, dsm, __u64)) {
+		printf("%s: io_args < sizeof_args), %d < %ld\n", __func__,
+				args->args_size, sizeof_args(struct nvme_io_args, dsm, __u64));
 		errno = EINVAL;
 		return -1;
 	}
+
+	if (args->args_size == sizeof_args(struct nvme_io_args, pif, __u64)) {
+		if (nvme_set_var_size_tags(&cmd.cdw2, &cmd.cdw3, &cmd.cdw14, args->pif,
+				args->sts, args->reftag_u64, args->storage_tag)) {
+			printf("set_var_size_tags failed\n");
+			errno = EINVAL;
+			return -1;
+		}
+	} else {
+		cmd.cdw2 = args->storage_tag & 0xffffffff;
+		cmd.cdw3 = (args->storage_tag >> 32) & 0xffff;
+		cmd.cdw14 = args->reftag;
+	}
+
 	return nvme_submit_io_passthru(args->fd, &cmd, args->result);
```

Comment:
Also the original changes were addressing a bug in the current handling of storage_tag, in that the spec defines both variable-sized tags as able to take the back 16 bits of cdw2, the full cdw3, and full cdw14. 

Calling it a "storage tag" feels a bit wrong when it could be entirely part of a larger reference tag, but in any case it shouldn't be setting the full bits of cdw2 and should be setting cdw3.

Reply:
Changed the code to the suggested way above.  Also fixed the original leg to be correct based on @bjpaupor  comment.